core/shared/platform/zephyr/zephyr_file.c

```diff
@@ -0,0 +1,975 @@
+// from the Zephyr POSIX configuration
+#define CONFIG_WASI_MAX_OPEN_FILES 16
+
+// Macro to retrieve a file system descriptor and check it's validity.
+#define GET_FILE_SYSTEM_DESCRIPTOR(fd, ptr)         \
+    do {                                            \
+        k_mutex_lock(&desc_array_mutex, K_FOREVER); \
+        ptr = &desc_array[(int)fd];                 \
+        if (!ptr) {                                 \
+            k_mutex_unlock(&desc_array_mutex);      \
+            return __WASI_EBADF;                    \
+        }                                           \
+        k_mutex_unlock(&desc_array_mutex);          \
+    } while (0)
+
+// Array to keep track of file system descriptors.
+static struct zephyr_fs_desc desc_array[CONFIG_WASI_MAX_OPEN_FILES];
+
+// mutex to protect the file descriptor array
+K_MUTEX_DEFINE(desc_array_mutex);
+
+static char prestat_dir[MAX_FILE_NAME + 1];
+
+static struct zephyr_fs_desc *
+zephyr_fs_alloc_obj(bool is_dir, const char *path, int *index)
+{
+    int i;
+    struct zephyr_fs_desc *ptr = NULL;
+    *index = -1; // give a default value to index in case table is full
+
+    k_mutex_lock(&desc_array_mutex, K_FOREVER);
+    for (i = 0; i < CONFIG_WASI_MAX_OPEN_FILES; i++) {
+        if (desc_array[i].used == false) {
+            ptr = &desc_array[i];
+            ptr->used = true;
+            ptr->is_dir = is_dir;
+            ptr->path = strdup(path);
+            if (ptr->path == NULL) {
+                ptr->used = false;
+                k_mutex_unlock(&desc_array_mutex);
+                return NULL;
+            }
+            *index = i;
+            break;
+        }
+    }
+
+    k_mutex_unlock(&desc_array_mutex);
+
+    return ptr;
+}
+
+static inline void
+zephyr_fs_free_obj(struct zephyr_fs_desc *ptr)
+{
+    free(ptr->path);
+    ptr->path = NULL;
+    ptr->used = false;
+}
+
+void
+debug_zephyr_fs_desc(const zephyr_fs_desc *desc)
+{
+    if (desc == NULL) {
+        os_printf("Descriptor is NULL\n");
+        return;
+    }
+    os_printf("Descriptor found at %p\n", desc);
+    os_printf("    Path: %s\n", desc->path ? desc->path : "NULL");
+    os_printf("    Is Directory: %s\n", desc->is_dir ? "Yes" : "No");
+    os_printf("    Used: %s\n", desc->used ? "Yes" : "No");
+
+    if (desc->is_dir) {
+        os_printf("    Directory: %p\n", desc->dir);
+    }
+    else {
+        os_printf("    File: %p\n", desc->file);
+    }
+}
+
+/* /!\ Needed for socket to work */
+__wasi_errno_t
+os_fstat(os_file_handle handle, struct __wasi_filestat_t *buf)
+{
+    struct zephyr_fs_desc *ptr = NULL;
+    int socktype, rc;
+
+    if (!handle->is_sock) {
+        GET_FILE_SYSTEM_DESCRIPTOR(handle->fd, ptr);
+        // debug_zephyr_fs_desc(ptr);
+
+        /* We treat the case of std[in/out/err] */
+        if (ptr->path != NULL
+            && (!strcmp(ptr->path, "stdin") || !strcmp(ptr->path, "stdout")
+                || !strcmp(ptr->path, "stderr"))) {
```

Comment:
I'm not sure because `ptr->path` is assigned in :
* `zephyr_fs_alloc_obj` 
* `os_renameat` 

In my opinion, it is really unlikely that an `open` (call `zephyr_fs_alloc_obj` ) and `rename` will be called concurently and even if it does happen it should be the user responsabilityto put a synchronisation mechanism at a higher level on a `FILE *` or `DIR*` to avoid concurrent access.

If I'm not clear enough please tell me.

Reply:
OK, so your opinion is to protect the slot of desc_array, but not to protect the accessing of the element in that slot, and protecting the latter is the user's responsibility. So let's leave it unchanged now, and check whether we need to do more in the future if there is issue reported?